active/0019-plugins.md

```diff
@@ -0,0 +1,151 @@
+
+Worth mentioning, since 4.3.2, EMQ X for the first supported drop-in installation of external plugins. That is, plugins can be compiled and packaged separately (instead of being released as a part of EMQ X package), see [Develop and deploy EMQ X plugin for Enterprise 4.3](https://emqx.atlassian.net/wiki/spaces/EMQX/blog/2021/05/23/168591472)
+
+## References
+
+- RabbitMQ: [Plugin Development Basics — RabbitMQ](https://www.rabbitmq.com/plugin-development.html)
+- Grafana: [Pie Chart](https://grafana.com/grafana/plugins/grafana-piechart-panel/)
+- WordPress: [WordPress Plugins](https://wordpress.org/plugins/)
+- RabbitMQ Plugin release packages: [Releases · rabbitmq/rabbitmq-management-exchange](https://github.com/rabbitmq/rabbitmq-management-exchange/releases)
+
+### High level requirements
+
+1. Hook points should be compatible with 4.x
+1. More clear ordering of all hook callbacks under one certain hook point, for example, internal hooks use priority 1000, and force external plugins to provide a priority number. If they wish to have it ordered before internal plugins they can choose to use a number smaller than 1000.
+1. All built-in features such as authentication and authorisation are not presented as plugins except for
+    1. LDAP authentication
+    1. PSK authentication
+
+### Basic steps to install an external plugin
+
+- Download compiled zip package
+- Upload to a specific dir
+- Execute a command to validate & install & enable & uninstall the plugin
+
+### Manage plugins from Dashboard UI
+
+- Manage installation from Dashboard GUI
+    - upload (and extract, but not persist it)
+    - install
+    - uninstall
+- Manage a list of installed plugins, supported actions:
+    - List view
+    - Show running status: "running" or "stopped".
+      Status should be presented per-node. e.g. `"status": "running"` for the current node (serving the API), or `"node_status": [{"node": "node1", "status": "running"}i, ...]` for a summary view of all nodes in the cluster.
+    - support actions: "start" or "stop"
+
+## Plugin metadata
+
+The plugin package should include metadata (in JSON format) to help identify, validate, or describe the package.
+
+- Name (same as the Erlang application name, it has to be globally unique)
+- Version
+- Author
+    - Name
+    - Contact
+- URL to source code
+- What functionalities (one or more of below)
+    - authentication
+    - authorisation
+    - data_persistence
+    - rule_engine_extension
+- Compatibility
+    - Compatible with EMQ X version(s), implicit low boundary of supported versions range is `5.0.0`, also to support version compares: `~>`, `>=`, `>`, `<=`, `<`, `==`
+      ref: https://github.com/erlang/rebar3/blob/c102379385013896711bba3969f280f851c67cc7/src/rebar_packages.erl#L376-L392
+    - Supported OTP releases (has to be the same as EMQ X's supported OTP versions)
```

Comment:
Code signature?

Reply:
do you mean read info from beam ?

Reply:
I mean something like GPG signature used to sign packages in linux.